pkg/promtail/targets/syslogparser/syslogparser.go

```diff
@@ -0,0 +1,71 @@
+package syslogparser
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+
+	"github.com/influxdata/go-syslog/v2"
+	"github.com/influxdata/go-syslog/v2/octetcounting"
+	"github.com/influxdata/go-syslog/v2/rfc5424"
+)
+
+// ParseStream parses a rfc5424 syslog stream from the given Reader, returning a channel
+// emitting the parsed messages. The parser automatically detects octet counting.
+// The channel is closed on EOF or unrecoverable errors.
+func ParseStream(r io.Reader) <-chan *syslog.Result {
+	results := make(chan *syslog.Result)
+	var err error
+	defer func() {
+		if err == nil {
+			return
+		}
+		go func(err error) {
+			results <- &syslog.Result{Error: err}
+			close(results)
+		}(err)
+	}()
+
+	buf := bufio.NewReader(r)
+	firstByte, err := buf.Peek(1)
+	if err != nil {
+		return results
+	}
+
+	b := firstByte[0]
+	if b == '<' {
+		go newlineSeparated(buf, results)
+	} else if b >= '0' && b <= '9' {
+		go octetCounting(buf, results)
+	} else {
+		err = fmt.Errorf("invalid or unsupported framing. first byte: '%s'", firstByte)
+	}
+
+	return results
+}
+
+func octetCounting(r io.Reader, results chan<- *syslog.Result) {
+	defer close(results)
+
+	callback := func(res *syslog.Result) {
+		results <- res
+	}
+
+	octetcounting.NewParser(syslog.WithListener(callback)).Parse(r)
+}
+
+func newlineSeparated(r io.Reader, results chan<- *syslog.Result) {
```

Comment:
Cannot this be replaced with `nontransparent.NewParser(syslog.WithListener(callback)).Parse(r)`?

Reply:
Yes, it should be possible. [Telegraf](https://github.com/influxdata/telegraf/blob/1f5be2bac7b895867411406b2eeac892b9ff9258/plugins/inputs/syslog/syslog.go#L322) uses it like this.

However its buffering is a bit strange. The line is emitted only after the next line is received, which is not optimal for connections with very few logs.

Reply:
OK. I wouldn't worry about it too much, hopefully nobody will use this mode, but let's keep it.